controllers/hypershiftlogforwarder/hypershiftlogforwarder_controller.go

```diff
@@ -120,23 +118,55 @@ func (r *HyperShiftLogForwarderReconciler) Reconcile(ctx context.Context, req ct
 		return ctrl.Result{}, err
 	}
 
-	deletion := false
-	// HyperShiftLogForwarder was deleted by CU
-	if !instance.DeletionTimestamp.IsZero() {
-		deletion = true
+	if !instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is being deleted
 		if controllerutil.ContainsFinalizer(instance, constants.ManagedLoggingFinalizer) {
-			controllerutil.RemoveFinalizer(instance, constants.ManagedLoggingFinalizer)
-			if err = r.Update(ctx, instance); err != nil {
+			hlfDeletion = true
+
+			//delete the clf if present
+			if clfFound {
+				err = r.MCClient.Delete(ctx, clf)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+			}
+
+			//create new clf from clft
+			if err := r.createClusterLogForwarderFromTemplate(ctx, req); err != nil {
 				return ctrl.Result{}, err
 			}
-		}
-	} else {
-		if !controllerutil.ContainsFinalizer(instance, constants.ManagedLoggingFinalizer) {
-			controllerutil.AddFinalizer(instance, constants.ManagedLoggingFinalizer)
-			if err = r.Update(ctx, instance); err != nil {
+
+			// remove our finalizer from the list and update it.
+			controllerutil.RemoveFinalizer(instance, constants.ManagedLoggingFinalizer)
+			if err := r.Update(ctx, instance); err != nil {
 				return ctrl.Result{}, err
 			}
 		}
+		r.log.V(1).Info("instance Info", "UID", instance.UID, "Name", instance.Name, "found", hlfFound, "deletion", hlfDeletion)
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+
+	}
+
+	// The object is not being deleted, so if it does not have our finalizer,
+	// then lets add the finalizer and update the object. This is equivalent
+	// registering our finalizer.
+	if !controllerutil.ContainsFinalizer(instance, constants.ManagedLoggingFinalizer) {
+		controllerutil.AddFinalizer(instance, constants.ManagedLoggingFinalizer)
+		if err := r.Update(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	r.log.V(1).Info("instance Info", "UID", instance.UID, "Name", instance.Name, "found", hlfFound, "deletion", hlfDeletion)
+
+	if req.NamespacedName.Name != constants.SingletonName {
```

Comment:
Why this is moved to bottom? It would be better if we can validate it earlier and stop if it is incorrect?

Reply:
Need to set the finalizers before updating the status , That'e hy this condition to the bottom